### PR TITLE
Universes

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR transitions the compiler to use **universes** instead of the **leak-check**. It is marked as [WIP] for a few reasons:

- The diagnostics at present are terrible =)
- This changes the behavior of coherence, regressing some things that used to compile 

The goals of this PR at present are:

- To start getting some eyes on the code
- To do a crater run
- To see the full travis results (due to https://github.com/rust-lang/rust/issues/52452, I am not able to run the full test suite locally anymore at present)

The first few commits in the PR are changing how `evaluate` treats regions. We now track whether region comparisons occurred, reverting the "staticized" query approach that @arielb1 put in. The problem with "staticized" queries is that it relied on the leak-check to get higher-ranked things correct, and we are removing the leak-check in this PR series, so that caused problems.

You can see at the end a collection of test updates. Mostly we behave the same but with atrocious diagnostics, but there are a number of cases where we used to error and now no longer do, as well as single case where we used to **not** error but we now do (the coherence-subtyping change).

(Note: it would be possible to do a version of leak-check that propagates universe information and recover the old behavior. I am reluctant to do so because I'd like to leave us room to get more precise -- e.g., I want to eventually handle things like `exists<'a> { for<'b> { if ('a: 'b) { 'a: 'b } } }` which presently the leak-check cannot cope with etc. Also because it seems more consistent to me: most folks I've talked to expect the new behavior and are surprised to learn that binding sites were so significant before when it comes to coherence. One question is, though, to what extent are people relying on this in the wild?)

### CODE REVIEW

**Comment by nikomatsakis** [comment]
(NB: This builds on https://github.com/rust-lang/rust/pull/55305)

**Comment by nikomatsakis** [comment]
@craterbot run start=master#f6e9a6e41cd9b1fb687e296b5a6d4c6ad399f862 end=try#dc8b02d6f1e93f212fc39844ea688b634b4a03c7 mode=check-only

**Review by tmandry — COMMENTED** [review]
left some (very minor) nits on tests

File: src/test/ui/coherence/coherence-subtyping.rs [hunk 1]
```diff
@@ -20,6 +19,7 @@ impl Contravariant for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
 }
 
 impl Contravariant for for<'a> fn(&'a u8, &'a u8) -> &'a u8 {
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/ui/coherence/coherence-subtyping.rs]: Uh... I think you're right. We used to have subtyping even in trait matching, and I think those cases date from that rule.

File: src/librustc/ty/util.rs [hunk 2]
```diff
@@ -861,11 +861,13 @@ fn is_copy_raw<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     let (param_env, ty) = query.into_parts();
     let trait_def_id = tcx.require_lang_item(lang_items::CopyTraitLangItem);
     tcx.infer_ctxt()
-       .enter(|infcx| traits::type_known_to_meet_bound(&infcx,
-                                                       param_env,
-                                                       ty,
-                                                       trait_def_id,
-                                                       DUMMY_SP))
+        .enter(|infcx| traits::type_known_to_meet_bound_modulo_regions(
```
Comment (eddyb) [inline src/librustc/ty/util.rs]: Should we advertise and/or track this behavior more strongly, or even consider fixing it?

I remember seeing something about MIR typeck/borrowck being able to take into account the fact that `Copy` impls can impose constraints on/between lifetimes.

Maybe we should be conservative and require that no lifetime constraints are needed to satisfy the `Copy` bound, for methods like these?
I guess I don't know what's using them anymore.

**Comment by nikomatsakis** [comment]
Crater results all look spurious.

File: src/librustc/ty/util.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/ty/util.rs]: Urgh, I'm not quite sure what you meant, @eddyb, when you said "fixing it". I guess you mean -- decide whether or not to consider the type to be copy? 

I am not sure if we can do that, nor if we want to. My goal is for the lowering to MIR is something to be done with erased regions, which implies we should be able to decide "copy or not" on that basis. Then we reconstruct regions entirely in the NLL pass (the NLL pass currently uses *nothing* from type check except for which regions are bound -- in my ideal world, we'd infer that *too*, but I've not put serious thought into it).

The NLL type checker however does enforce the full set of constraints about regions when you try to copy a type.

Comment (eddyb) [reply src/librustc/ty/util.rs]: Sure, what I mean is that these e.g. `Ty::is_copy` methods are far too general and will get misused if we leave them like this.
So I would prefer if this sort of thing was more opt-in at the call sites, if possible.

Comment (nikomatsakis) [reply src/librustc/ty/util.rs]: I will rename them for now to `is_copy_modulo_regions` or something

File: src/librustc/ty/util.rs [hunk 3]
```diff
@@ -639,13 +646,24 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
...
                     tcx_at: TyCtxtAt<'a, 'tcx, 'tcx>,
                     param_env: ty::ParamEnv<'tcx>)-> bool
     {
         tcx_at.is_sized_raw(param_env.and(self))
     }
 
+    /// Checks whether values of this type `T` implement the `Freeze`
+    /// trait -- frozen types are those that do not contain a
+    /// `UnsafeCell` anywhere.  This is a language concept used to
+    /// determine how to handle `static` values, the trait itself is
```
Comment (eddyb) [inline src/librustc/ty/util.rs]: Not just `static`s but also the semantics of `&T` (e.g. we tell LLVM `&T` is read-only iff `T: Freeze`). In general, whether shared values are truly immutable, or interior mutability is allowed.

**Comment by eddyb** [comment]
Should we wait for https://github.com/rust-lang/rust/pull/55517#discussion_r234671373?

**Comment by nikomatsakis** [comment]
I believe that the prior problems were fixed by the move to make autoderef a query. 

@bors r=scalexm

**Comment by nikomatsakis** [comment]
@matthiaskrgr are you working on fixing the toolstate change here? (I see some commits that look suggestive)

**Comment by matthiaskrgr** [comment]
Yeah, sorry about the spam.

There is a recent regression in clippy https://github.com/rust-lang/rust-clippy/issues/3628 which has not been very easy to track down which prevents use from landing the fix. :/

**Pull request merged**